CMakeLists.txt

```diff
@@ -419,6 +419,15 @@ endif()
 #
 if (${CMAKE_SYSTEM_NAME} MATCHES "Linux")
     configure_files (${CMAKE_CURRENT_SOURCE_DIR}/dist/rpm ${CMAKE_BINARY_DIR}/rpm)
+    if (DEFINED ENV{SNAPCRAFT_PART_BUILD})
+        set (SYSTEMD_EXEC_PATH_PREFIX "/snap/bin/barrier.")
+    else()
+        set (SYSTEMD_EXEC_PATH_PREFIX "/usr/bin/")
+    endif()
+    configure_files (${CMAKE_CURRENT_SOURCE_DIR}/dist/systemd ${CMAKE_BINARY_DIR}/systemd)
+    install(FILES ${CMAKE_BINARY_DIR}/systemd/barrierc@.service DESTINATION lib/systemd/system)
+    install(FILES ${CMAKE_BINARY_DIR}/systemd/barriers@.service DESTINATION lib/systemd/system)
+    install(FILES ${CMAKE_BINARY_DIR}/systemd/barriers.service DESTINATION lib/systemd/system)
```

Comment:
No idea how to do this in cmake but `pkg-config --variable systemd_system_unit_dir systemd` is the best way to get this, straight from the horse's mouth. Requires `systemd.pc` to be installed.


But that aside - wouldn't installing this as user service be a better option? To run X and thus barrier you need a logged-in user anyway. And at that point you can enable/disable it per user and the service has access to the user's home directory (and thus config files, etc.). This goes doubly for the barrierc service.

Reply:
I would tend to agree, but equally we can't assume `systemd.pc` is installed? I prefer `--user` services, but some people have said they want Barrier to load before login. This just isn't feasible. Personally, I'd rather go for `--user`...

Reply:
>  we can't assume systemd.pc is installed?

Two options here: either require the dependency (like all others we already require anyway) or default to pkgconfig and if not available pick a fallback path. Since this is unlikely to ever change much anyway, that's a fairly safe option too.

As for user vs system: maybe a configure-time build option?

Reply:
Second option feels more sane to me. I think option for user vs. system might not be suited for configure-time - bear in mind other distros, if not already, will be packaging Input Leap, wouldn't that result in more packages for user vs. system approaches? Or am I reading this wrong?

Reply:
depends mostly on packaging, I guess. But what you could also do is simply install both. pipewire for example does this via the meson options `-Dsystemd-system-service=enabled` (disabled by default) and `-Dsystemd-user-service=enabled` (enabled by default).

Also, just because they're installed doesn't mean they're *enabled* (`systemctl enable --now input-leap` or so) and what is enabled by default is a distribution-specific toggle anyway.

Reply:
Yeah, let's just install user services for now, as system services are a bit more hacky, and require the IL config to be stored somewhere, which AFAIK the approach is user-dependent. I can't see system services working for IL just yet. It's something we can explore.

Reply:
Hey folks, I wanted to suggest writing files to `/etc/xdg/autostart` , because it seems that [arch](https://wiki.archlinux.org/title/XDG_Autostart) and [sway (a wayland i3 clone)](https://github.com/swaywm/sway/issues/1423) push people that way. With a sort of your desktop manager should help manage session autostarting, but if not, just use [dex](https://github.com/jceb/dex) mentality. For background some of the consideration into this suggestion.

It looks like [systemd wasn't really designed for a per seat/per session](https://lists.freedesktop.org/archives/systemd-devel/2014-March/017552.html) service. I stumbled on the above from the Note at the bottom of [Arch's systemd's How it works](https://wiki.archlinux.org/title/systemd/User#how_it_works), which explains that it's more a per user than per session thing.

I see the `Environment=DISPLAY=:0` which in fighting with nvidia and gdm3 today, I discovered that might [not be true for a good amount of environments](https://askubuntu.com/a/1323971).

I considered but did not want to suggest just writing to [.xprofile](https://wiki.archlinux.org/title/Xprofile), as that doesn't seem to fit with upcoming wayland.

P.S. Thanks for all the awesome work so far. I've built the project on my linux boxes and just need to do the same on windows (the depends are a bit bigger) to replace synergy. Also, I'd be happy to whip up a PR if you all are interested in using `.desktop` files.

Reply:
> I consider systemd/user as a good alternative for dex's autostart functionality and switched to it recently

that's from the [dex README](https://github.com/jceb/dex) itself :) So that's not exactly a huge endorsement of using autostart + dex.

Reply:
to me the users complaining about systemd user session not allowing to use IL before login means before login as a user.
The graphical login manager also runs as a user so this graphical login manager could have its own systemd user session service instance for IL and we would have the IL at boot.
The poster seems to not understand the graphical login manager is itself logged in as its own user, only automatically. I do not believe they requested IL to be working before the graphical login manager.


About per seat:
Also, in this patchset, the system-wide systemd service would not help with multi-seat either. Either way, I see no shortcomings in binding IL to the user session instead of the login session.
However, I agree that IL should be seat-aware in that we should have different screen mapping depending on your seat (ie you could have boxes with one display on one seat, three on another, two with two displays, etc). IL config syntax does not support multi-seat as far as I know.

In https://lists.freedesktop.org/archives/systemd-devel/2014-March/017552.html the poster confused login session to seat (he says about login session then says he meant different window managers, so I believe he meant different graphical logins). Lennart discards his request because in his understanding graphical display managers will be per user so he sees no point in the request (as systemd already has per-user session support). It is even more obvious in https://lists.freedesktop.org/archives/systemd-devel/2014-March/017543.html where Lennart says that the per login management should be the window manager job, not systemd one. Also, from [this email](https://lists.freedesktop.org/archives/systemd-devel/2014-March/017552.html), I understand that Lennart sees the per-user session display manager has the arbitrator for the devices (including virtual devices like IL input I assume).

This to me means if we want per seat, IL could become part of the display manager or kind of a plugin to it.
Or maybe we could have code in the display manager that spawns systemd user per instance services (one for each seat). That is have one IL server and clients set for each seat on the boxes.
Note that systemd-logind is aware of the seat setup. But as far as I know, we cannot have it spawn programs on login based on seat.

Maybe a bug report for multi-seat support should be opened in input-leap. I believe this to be a feature request which should not prevent this pull request from merging.
The issue already exists with current code. And would not be made worse by systemd user services for IL.